cpp/src/arrow/dataset/discovery.h

```diff
@@ -85,35 +85,97 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
   std::shared_ptr<Expression> root_partition_;
 };
 
-/// \brief FileSystemDataSourceFactory creates a DataSource from a vector
-/// of fs::FileStats or a fs::Selector.
+struct FileSystemDiscoveryOptions {
+  // For the purposes of applying the partition scheme, paths will be stripped
+  // of the partition_base_dir. Files not matching the partition_base_dir
+  // prefix will be skipped for partition discovery. The ignored files will still
+  // be part of the DataSource, but will not have partition information.
+  //
+  // Example:
+  // partition_base_dir = "/dataset";
+  //
+  // - "/dataset/US/sales.csv" -> "US/sales.csv" will be given to the partition
+  //                              scheme
+  //
+  // - "/home/john/late_sales.csv" -> Will be ignored for partition discovery.
+  //
+  // This is useful for partition schemes which parses directory when ordering
+  // is important, e.g. SchemaPartitionScheme.
+  std::string partition_base_dir;
+
+  // Files given (via selector or explicitly) will be filtered via the
+  // FileFormat::IsSupported method.  This will incur IO for each files in a
+  // serial and single threaded fashion. Disabling this feature will skip the
+  // IO, but unsupported files may will be present in the DataSource
+  // (resulting in an error at scan time).
+  bool filter_supported_files = true;
```

Comment:
This param name isn't clear to me. Just brainstorming, but `exclude_invalid_files`, `check_file_validity`, etc. are more explicit about what will happen. The latter might be better because it makes clearer that work will be done, which you may want to skip.